File: databinding/src/main/java/io/stanwood/framework/databinding/adapters/ViewPagerBindingAdapter.java
package io.stanwood.framework.databinding.adapters;

import androidx.databinding.BindingAdapter;
import androidx.databinding.adapters.ListenerUtil;
import androidx.viewpager.widget.ViewPager;
import io.stanwood.framework.databinding.R;

public class ViewPagerBindingAdapter {

    @BindingAdapter("onPageScrolled")
    public static void setListener(ViewPager view, OnPageScrolled pageScrolled) {
        setListener(view, pageScrolled, null, null);
    }

    @BindingAdapter("onPageSelected")
    public static void setListener(ViewPager view, OnPageSelected pageSelected) {
        setListener(view, null, pageSelected, null);
    }

    @BindingAdapter("onPageScrollStateChanged")
    public static void setListener(ViewPager view, OnPageScrollStateChanged scrollStateChanged) {
        setListener(view, null, null, scrollStateChanged);
    }

    @BindingAdapter({"onPageScrolled", "onPageSelected", "onPageScrollStateChanged"})
    public static void setListener(ViewPager view, final OnPageScrolled pageScrolled,
                                   final OnPageSelected pageSelected, final OnPageScrollStateChanged pageScrollStateChanged) {
        final ViewPager.OnPageChangeListener newListener;
        if (pageScrolled == null && pageSelected == null && pageScrollStateChanged == null) {
            newListener = null;
        } else {
            newListener = new ViewPager.OnPageChangeListener() {
                @Override
                public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                    if (pageScrolled != null) {
                        pageScrolled.onPageScrolled(position, positionOffset, positionOffsetPixels);
                    }
                }

                @Override
                public void onPageSelected(int position) {
                    if (pageSelected != null) {
                        pageSelected.onPageSelected(position);
                    }
                }

                @Override
                public void onPageScrollStateChanged(int state) {
                    if (pageScrollStateChanged != null) {
                        pageScrollStateChanged.onPageScrollStateChanged(state);
                    }

                }
            };
        }
        final ViewPager.OnPageChangeListener oldListener = ListenerUtil.trackListener(view,
                newListener, R.id.onPageChangeListener);
        if (oldListener != null) {
            view.removeOnPageChangeListener(oldListener);
        }
        if (newListener != null) {
            view.addOnPageChangeListener(newListener);
        }

    }

    public interface OnPageScrolled {
        void onPageScrolled(int position, float positionOffset, int positionOffsetPixels);
    }

    public interface OnPageSelected {
        void onPageSelected(int position);
    }

    public interface OnPageScrollStateChanged {
        void onPageScrollStateChanged(int state);
    }

}

File: framework-databinding/src/main/java/io/stanwood/framework/databinding/adapters/ViewPagerBindingAdapter.kt
package io.stanwood.framework.databinding.adapters

import androidx.databinding.BindingAdapter
import androidx.databinding.adapters.ListenerUtil
import androidx.viewpager.widget.ViewPager
import io.stanwood.framework.databinding.R

object ViewPagerBindingAdapter {

    @JvmStatic
    @BindingAdapter("onPageScrolled")
    fun setListener(view: ViewPager, pageScrolled: OnPageScrolled?) {
        setListener(view, pageScrolled, null, null)
    }

    @JvmStatic
    @BindingAdapter("onPageSelected")
    fun setListener(view: ViewPager, pageSelected: OnPageSelected?) {
        setListener(view, null, pageSelected, null)
    }

    @JvmStatic
    @BindingAdapter("onPageScrollStateChanged")
    fun setListener(view: ViewPager, scrollStateChanged: OnPageScrollStateChanged?) {
        setListener(view, null, null, scrollStateChanged)
    }

    @JvmStatic
    @BindingAdapter("onPageScrolled", "onPageSelected", "onPageScrollStateChanged")
    fun setListener(
        view: ViewPager,
        pageScrolled: OnPageScrolled?,
        pageSelected: OnPageSelected?,
        pageScrollStateChanged: OnPageScrollStateChanged?
    ) {
        val newListener: ViewPager.OnPageChangeListener?
        if (pageScrolled == null && pageSelected == null && pageScrollStateChanged == null) {
            newListener = null
        } else {
            newListener = object : ViewPager.OnPageChangeListener {
                override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                    pageScrolled?.onPageScrolled(position, positionOffset, positionOffsetPixels)
                }

                override fun onPageSelected(position: Int) {
                    pageSelected?.onPageSelected(position)
                }

                override fun onPageScrollStateChanged(state: Int) {
                    pageScrollStateChanged?.onPageScrollStateChanged(state)

                }
            }
        }
        ListenerUtil.trackListener<ViewPager.OnPageChangeListener>(view, newListener, R.id.onPageChangeListener)
            ?.apply {
                view.removeOnPageChangeListener(this)
            }
        newListener?.apply {
            view.addOnPageChangeListener(this)
        }
    }

    interface OnPageScrolled {
        fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int)
    }

    interface OnPageSelected {
        fun onPageSelected(position: Int)
    }

    interface OnPageScrollStateChanged {
        fun onPageScrollStateChanged(state: Int)
    }
}